U01: Create lobby screen and basic client structure

paul891112 opened the pull request:
Implement basic structure of lobby

Modify SessionManager, PlayerManager, create template for TileManager

No navigation from titlescreen to lobbyscreen implemented yet

To view lobbyscreen, in src/core/main.tsx set scene: [LobbyScreen]

To simulate client-server interaction, mock methods are defined

As a user, I can now see how people join the lobby.

I built it on a side branch because master is missing key components like managers and events

Are we going to use custom designed button? Put a basic button and text on top? Or just use text as button?

C0DECYCLE reviewed (changes requested):
Thanks a lot! Will review it this weekend. Could you please in the meantime adjust the pr title so that its in the following pattern: "Userstory: PR/Issue description"… we forgot to add it to the readme but will do so this weekend.

Edit: You can look at the other prs for sample titles.

paul891112 commented:
> Thanks a lot! Will review it this weekend. Could you please in the meantime adjust the pr title so that its in the following pattern: "Userstory: PR/Issue description"… we forgot to add it to the readme but will do so this weekend.
> 
> Edit: You can look at the other prs for sample titles. Ah and please rebase your branch onto master so that there is only your commit.

I changed the PR title and rebased the branch to master, hope that worked

paul891112 commented:
UI design questions in lobby:

Add restriction to length of playername?

What kind of button?

Start, share and quit buttons?

Any creative ideas?

C0DECYCLE commented:
> I changed the PR title and rebased the branch to master, hope that worked

Thanks for the PR title! The commits don't seem to have worked🤔 In the commit tab our old master commits are there as well. Maybe try that: 
1. Reset the local version of your branch to the current master. (DONT PUSH, OR YOUR COMMITS WILL BE LOST)
2. Then cherry pick your made commits from the remote version of your branch onto your local one. 
3. After that your local branch should only differ from main in your contributed commits and you can force push to remote and this pr will update.

C0DECYCLE commented:
> Are we going to use custom designed button? Put a basic button and text on top? Or just use text as button?

We will use custom phaser "buttons" with an image, i already showed @Croxsy how to do it... maybe ask her. She also has some needed util code.

C0DECYCLE commented:
> UI design questions in lobby:
> 
> Add restriction to length of playername?
> 
> What kind of button?
> 
> Start, share and quit buttons?
> 
> Any creative ideas?

Yess... 
- Would restrict the length of playername to something like 12-18 chars? 
- Button: A pixelart button image which made interactable via the util function @Croxsy already posses, but i can also provide.

Patertuck commented:
Great progress @paul891112 , keep it up!

C0DECYCLE commented:
> Great progress @paul891112 , keep it up!

+1 absolutely!

C0DECYCLE commented on src/managers/PlayerManager.ts:
```diff
@@ -1,20 +1,41 @@
 
 class PlayerManager {
     public readonly onSync: EventEmitter;
+    private playernames: Array<string> = [];
```
please initialize in constructor

C0DECYCLE commented on src/managers/PlayerManager.ts:
```diff
@@ -1,20 +1,41 @@
         });
     }
+
+    public getPlayers(){
```
add return type

C0DECYCLE commented on src/managers/PlayerManager.ts:
```diff
@@ -1,20 +1,41 @@
+
+            //for testing purpose
+            log("Update playername: " + name);
+
```
Please mock in seperate method as in sessionmanager.

C0DECYCLE commented on src/scenes/lobbyscreen.ts:
```diff
@@ -0,0 +1,148 @@
+        this.load.image("testButton", "assets/testButton.png");
+    }
+
+    graphics!: Phaser.GameObjects.Graphics;
```
put all fields above the methods and with access modifier

C0DECYCLE commented on src/scenes/lobbyscreen.ts:
```diff
@@ -0,0 +1,148 @@
+                fontStyle: fontstyle,                
+            });
+        }
+    }
```
Instead of update frame driven please make it event driven. so that the lobby information doesnt update each frame but update only when the inforamtion changes and then i think you can also leave away the private me and players field

C0DECYCLE replied:
Looks much better thanks!

C0DECYCLE commented on src/scenes/lobbyscreen.ts:
```diff
@@ -0,0 +1,148 @@
+            this.game.canvas.style.cursor = "default";
+        });
+
+    }
```
please add variable return type everywhere and get the interactivity function from @Croxsy and use this to make things interactive.

paul891112 replied:
Cant add the type "string" to this.game.canvas.style.cursor without TypeScript complaining, it should be a string that holds a CSS value.

C0DECYCLE reviewed (changes requested):
Thanks for the great work! First review wave is finally in, sorry for the wait.

PS: Please remove the unused assets (bomb, dude, platform, sky, star)

paul891112 commented on src/managers/SessionManager.ts:
```diff
@@ -1,13 +1,17 @@
 class SessionManager {
     public readonly onSync: EventEmitter;
 
+    private information: Array<string|int> = [];
```
I assume this change is not necessary, since now we have private information: SessionInformation = {
sessionId: UUID = "",
... = "",
... = -1, 
...
}
meaning this.information upon initialization contains default values of type SessionInformation.

paul891112 commented:
First requested changes are made, now call to update() is event-based, other changes are also completed
Add @Croxsy implementation of interactify()
I changed the definition and usage alittle such that cursor changes to pointer (finder) when hover over a button.

C0DECYCLE commented:
Amazing thank you very much! Will rereview it when i have time. Is it ok when we then merge  the @Croxsy PR first because of the util files.

paul891112 commented:
Sure no problem, might have to change manager structure anyway

C0DECYCLE commented on src/scenes/lobbyscreen.ts:
```diff
@@ -0,0 +1,99 @@
+        this.update();
+    }
+
+    public syncMe = () => {
```
same here and above please with the arrow syntax 

(and return type)

C0DECYCLE commented on src/scenes/lobbyscreen.ts:
```diff
@@ -0,0 +1,99 @@
+        this.graphics.strokeRoundedRect(100, ypos, 600, 27, 5);
+    }
+
+    public override update(): void {
```
attention the method called update is phaser reserved and will be called every frame... you need to rename it to something else.

C0DECYCLE commented on src/scenes/lobbyscreen.ts:
```diff
@@ -0,0 +1,99 @@
+                stroke: "#000000",
+                strokeThickness: 3,
+                fontStyle: fontstyle,                
+            });
```
it looks to me that every update all new texts and background get created and not properly destroyed, creating a memory leak.

paul891112 commented:
Implement InformationTypes
Create methods to compare two players, two tiles
Implement GeneralManager that constantly requests data from server
Make SessionManager and PlayerManager update their data storage if changes are detected
Define some variables as Nullable<...> since some attributes can be null
Create functions for buttons (implement invite)
Use mock functions to test functionality, official version is in comments
fetchdata() in GeneralManager is in comment (in constructor)
Fix second PR review
Now lobby screen clears frame before updating it
Notes moving on:
SessionManager needs better testing with create lobby function to ensure robustness
TileManager still needs to be implemented

C0DECYCLE commented:
> Implement InformationTypes Create methods to compare two players, two tiles Implement GeneralManager that constantly requests data from server Make SessionManager and PlayerManager update their data storage if changes are detected Define some variables as Nullable<...> since some attributes can be null Create functions for buttons (implement invite) Use mock functions to test functionality, official version is in comments fetchdata() in GeneralManager is in comment (in constructor) Fix second PR review Now lobby screen clears frame before updating it Notes moving on: SessionManager needs better testing with create lobby function to ensure robustness TileManager still needs to be implemented

Amazing thank you really much for the great work! Im currently looking at it.

PS: For the future: its oke if you resolve fixed comments.

C0DECYCLE reviewed (approved):
LGTM
